### PR TITLE
(GH-811) Fix folding with spanning regions

### LINKED ISSUE #811: Code Folding not working with PoshBot code
Originally from https://github.com/PowerShell/vscode-powershell/issues/1631

---

## System Details
System Details Output
```
### VSCode version: 1.29.1 bc24f98b5f70467bc689abf41cc5550ca637088e x64

### VSCode extensions:
CoenraadS.bracket-pair-colorizer@1.0.61
DavidAnson.vscode-markdownlint@0.21.0
eamodio.gitlens@9.0.2
EditorConfig.EditorConfig@0.12.5
eg2.tslint@1.0.40
Equinusocio.vsc-material-theme@2.6.2
kisstkondoros.vscode-codemetrics@1.17.4
ms-vscode.csharp@1.17.1
ms-vscode.PowerShell@1.10.0
wesbos.theme-cobalt2@2.1.6


### PSES version: 1.10.0.0

### PowerShell version:

Name                           Value
----                           -----
PSVersion                      5.1.17763.134
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17763.134
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

## Issue Description
I am experiencing a problem with code folding. The SendMessage functions within the PoshBot Backends for slack and teams both fold rather weirdly.

Raw files:
https://raw.githubusercontent.com/poshbotio/PoshBot/master/PoshBot/Implementations/Teams/TeamsBackend.ps1
https://raw.githubusercontent.com/poshbotio/PoshBot/master/PoshBot/Implementations/Slack/SlackBackend.ps1

## Expected Behaviour
It should fold the entire function.

## Actual Behaviour
It folds parts of the function.

### PR DESCRIPTION
**Pull request by glennsarti**
From https://github.com/PowerShell/vscode-powershell/issues/1631

Fixes #811

---

In the use case above the following code snippet is confusing the folder;

Bad folder
``` powershell
foreach ($1 in $2) {     <----- STARTS MATCH HERE (1)

    $x = @{  <-----   STARTS MATCH HERE (2)
        'abc' = 'def'
    }        <----- ENDS   MATCH HERE (1) (2)
}
```

Good folder
``` powershell
foreach ($1 in $2) {     <----- STARTS MATCH HERE (1)

    $x = @{  <-----   STARTS MATCH HERE (2)
        'abc' = 'def'
    }        <----- ENDS  MATCH HERE (2)
}        <----- ENDS  MATCH HERE (1)
```

The old grammar based folder knew the difference between a `}` terminating a hash vs a script block.  The PowerShell Tokeniser does not.

Note I found this is also the case for  `@( ... )`, `( ... )`, and `$( ... )`

### CODE REVIEW

**Comment by glennsarti** [comment]
@rjmholt @TylerLeonhardt @rkeithhill Although this is WIP, looking for a quick review.  I need to cleanup the commit message.

File: src/PowerShellEditorServices/Language/TokenOperations.cs [hunk 1]
```diff
@@ -146,19 +140,19 @@ internal static class TokenOperations
         }
 
         /// <summary>
-        /// Given an array of tokens, find matching regions which start and end with a different TokenKind
+        /// Given an array of tokens, find matching regions which start (array of tokens) and end with a different TokenKind
         /// </summary>
         static private List<FoldingReference> MatchTokenElements(
             Token[] tokens,
-            TokenKind startTokenKind,
+            TokenKind[] startTokenKind,
```
Comment (TylerLeonhardt) [inline src/PowerShellEditorServices/Language/TokenOperations.cs]: I thought about this using a HastSet instead since it's faster to look up than iterating through an array but we're talking about 3 elements so I'm not too concerned.

File: test/PowerShellEditorServices.Test/Language/TokenOperationsTests.cs [hunk 2]
```diff
@@ -217,5 +217,33 @@ public class TokenOperationsTests
             FoldingReference[] result = GetRegions(testString);
             AssertFoldingReferenceArrays(expectedFolds, result);
         }
+
+        // This tests that token matching { -> }, @{ -> } and
+        // ( -> ), @( -> ) and $( -> ) does not confuse the folder
+        [Fact]
+        public void LaguageServiceFindsFoldablRegionsWithSameEndToken() {
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply test/PowerShellEditorServices.Test/Language/TokenOperationsTests.cs]: Are you talking about;

`Some-Function @splatme` ?

In that case the splat can't span lines, so no folding required.

File: src/PowerShellEditorServices/Language/TokenOperations.cs [hunk 3]
```diff
@@ -28,24 +29,17 @@ internal static class TokenOperations
...
-            );
-
-            // Find matching braces ( -> )
-            foldableRegions.AddRange(
-                MatchTokenElements(tokens, TokenKind.LParen, TokenKind.RParen, RegionKindNone)
-            );
-
-            // Find matching arrays @( -> )
+            // Find matching braces  { -> }
+            // Find matching hashes @{ -> }
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Is there a practical example of this?

I'll run this through the parser and see what I get

[the remark this answers is not in the document]
Comment (glennsarti) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Found and added as additional commit.

**Comment by glennsarti** [comment]
@TylerLeonhardt This is no longer WIP.  Ready for merge.

File: test/PowerShellEditorServices.Test/Language/TokenOperationsTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (glennsarti) [reply test/PowerShellEditorServices.Test/Language/TokenOperationsTests.cs]: I see... you want a negative test.  Done.

File: src/PowerShellEditorServices/Language/TokenOperations.cs [hunk 4]
```diff
@@ -28,29 +29,27 @@ internal static class TokenOperations
...
-            foldableRegions.AddRange(
-                MatchTokenElements(tokens, TokenKind.LCurly, TokenKind.RCurly, RegionKindNone)
-            );
-
-            // Find matching braces ( -> )
+            // Find matching braces  { -> }
+            // Find matching hashes @{ -> }
             foldableRegions.AddRange(
-                MatchTokenElements(tokens, TokenKind.LParen, TokenKind.RParen, RegionKindNone)
+                MatchTokenElements(tokens, new TokenKind[] { TokenKind.LCurly, TokenKind.AtCurly }, TokenKind.RCurly, RegionKindNone)
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Language/TokenOperations.cs]: Could this array be created in a static readonly field?

File: src/PowerShellEditorServices/Language/TokenOperations.cs [hunk 5]
```diff
@@ -28,29 +29,27 @@ internal static class TokenOperations
...
             );
 
-            // Find matching arrays @( -> )
+            // Find matching parentheses     ( -> )
+            // Find matching array literals @( -> )
+            // Find matching subexpressions $( -> )
             foldableRegions.AddRange(
-                MatchTokenElements(tokens, TokenKind.AtParen, TokenKind.RParen, RegionKindNone)
+                MatchTokenElements(tokens, new TokenKind[] { TokenKind.LParen, TokenKind.AtParen, TokenKind.DollarParen }, TokenKind.RParen, RegionKindNone)
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Language/TokenOperations.cs]: echo: Static readonly field for array

File: src/PowerShellEditorServices/Language/TokenOperations.cs (hunk 4 above)
Comment (glennsarti) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Yeah it could be

File: src/PowerShellEditorServices/Language/TokenOperations.cs (hunk 5 above)
Comment (glennsarti) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Yeah it could be

**Comment by glennsarti** [comment]
@SeeminglyScience Just a note that I can't reply easily to review comments.

> That way you'd only need to enumerate once, it should already be sorted, and no nulls. Sorry if this was already brought up in the original PR

* The sorting will not be solved due to the comments foldingreferences being appended later.
* The way it's currently written nulls can still be written if the range doesn't span at least two lines

Note that PR #806 removes all those code paths anyway.

File: src/PowerShellEditorServices/Language/TokenOperations.cs (hunk 4 above)
Comment (rjmholt) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Given how often we'll be folding, it's probably a good idea -- want to avoid GC pressure if we can. Should just be a matter of putting the following at the top of the class:

```csharp
// Opening tokens for { } and @{ }
private static readonly TokenKind[] s_openingBraces = new [] { TokenKind.LCurly, TokenKind.AtCurly };

// Opening tokens for ( ), @( ) and $( )
private static readonly TokenKind[] s_openingParens = new [] { TokenKind.LParen, TokenKind.AtParen, TokenKind.DollarParen };
```

File: src/PowerShellEditorServices/Language/TokenOperations.cs [hunk 6]
```diff
@@ -28,29 +29,27 @@ internal static class TokenOperations
...
 
-            // Find matching hashes @{ -> }
+            // Find contiguous here strings @' -> '@
             foldableRegions.AddRange(
-                MatchTokenElements(tokens, TokenKind.AtCurly, TokenKind.RParen, RegionKindNone)
+                MatchTokenElement(tokens, TokenKind.HereStringLiteral, RegionKindNone)
             );
 
-            // Find contiguous here strings @' -> '@
+            // Find unoppinionated variable names ${ \n \n }
```
Comment (rjmholt) [inline src/PowerShellEditorServices/Language/TokenOperations.cs]:
```suggestion
            // Find unopinionated variable names ${ \n \n }
```

File: test/PowerShellEditorServices.Test/Language/TokenOperationsTests.cs [hunk 7]
```diff
@@ -111,23 +114,30 @@ public class TokenOperationsTests
 # Comment Block 2
 # Comment Block 2
 $something = $true
-#endregion Comment Block 3";
+#endregion Comment Block 3
+
+# What about anonynous variable assignment
+${this
```
Comment (rjmholt) [inline test/PowerShellEditorServices.Test/Language/TokenOperationsTests.cs]: This is the edge case I like to see. Allowing these awful variable names is one of @JamesWTruher's proudest additions to PowerShell :)

File: src/PowerShellEditorServices/Language/TokenOperations.cs (hunk 4 above)
Comment (glennsarti) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Agreed. This codepath is removed in PR #806 and we use the AST

File: src/PowerShellEditorServices/Language/TokenOperations.cs (hunk 6 above)
Comment (glennsarti) [reply src/PowerShellEditorServices/Language/TokenOperations.cs]: Fixed

File: src/PowerShellEditorServices/Language/TokenOperations.cs [hunk 8]
```diff
@@ -16,10 +16,26 @@ namespace Microsoft.PowerShell.EditorServices
     /// </summary>
     internal static class TokenOperations
     {
+        // Region kinds to align with VSCode's region kinds
```
Comment (glennsarti) [inline src/PowerShellEditorServices/Language/TokenOperations.cs]: Thanks for adding that
